index.js

```diff
@@ -55,22 +74,60 @@ async function init() {
 
 	console.log('Source:', {repo, ref, dir});
 
+	if (!navigator.onLine) {
+		updateStatus('⚠ You are offline.');
+		throw new Error('User agent is offline');
+	}
+
 	updateStatus('Retrieving directory info…');
 
+	await ensureRepoIsAccessible(repo);
+
 	const files = await listContent.viaTreesApi(`${repo}#${ref}`, decodeURIComponent(dir), localStorage.token, ref);
 
-	updateStatus(`Downloading (0/${files.length}) files…`, '\n• ' + files.join('\n• '));
+	if (files.length > 0) {
+		updateStatus(`Downloading (0/${files.length}) files…`, '\n• ' + files.join('\n• '));
+	} else {
+		updateStatus('No files to download');
+		return;
+	}
 
 	let downloaded = 0;
-	const requests = await Promise.all(files.map(async path => {
-		const response = await fetch(`https://raw.githubusercontent.com/${repo}/${ref}/${path}`);
-		const blob = await response.blob();
+	let requests;
+	const controller = new AbortController();
+	try {
+		requests = await Promise.all(files.map(async path => {
+			const response = await fetch(
+				`https://raw.githubusercontent.com/${repo}/${ref}/${path}`,
+				{signal: controller.signal}
+			);
+
+			if (response.status >= 400) {
+				controller.abort();
+				return;
+			}
+
+			const blob = await response.blob();
+
+			downloaded++;
+			updateStatus(`Downloading (${downloaded}/${files.length}) files…`, path);
+
+			return {path, blob};
+		}));
+	} catch (error) {
+		controller.abort();
+
+		if (navigator.onLine === false) {
+			updateStatus('⚠ Could not download all files, network connection lost.');
+		} else if (error instanceof DOMException) {
+			updateStatus('⚠ Could not download all files.');
+		} else {
+			updateStatus('⚠ Some files were blocked from downloading, try to disable any ad blockers and refresh the page.');
```

Comment:
It might be worth specifying which file failed, at least in the console.

Reply:
I thought so, but it would need to reintroduce the weird `abortReason` construct. My reasoning was that devtools users could just find the error in the network tab, but that would at least require a refresh, so... yeah, this sounds reasonable to introduce.